include/rmm/mr/device/failure_alternate_resource_adaptor.hpp

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/detail/error.hpp>
+#include <rmm/detail/export.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/resource_ref.hpp>
+
+#include <cstddef>
+#include <mutex>
+#include <unordered_set>
+
+namespace RMM_NAMESPACE {
+namespace mr {
+/**
+ * @addtogroup device_resource_adaptors
+ * @{
+ * @file
+ */
+
+/**
+ * @brief A device memory resource that use an alternate upstream resource when the primary throw a
+ * specified exception type.
+ *
+ * An instance of this resource must be constructed with two existing upstream resource to
+ * satisfy allocation requests.
+ *
+ * @tparam PrimaryUpstream The type of the primary upstream resource used for
+ * allocation/deallocation.
+ * @tparam AlternateUpstream The type of the alternate upstream resource used for
+ * allocation/deallocation when the primary fails.
+ * @tparam ExceptionType The type of exception that this adaptor should respond to.
+ */
+template <typename PrimaryUpstream,
+          typename AlternateUpstream,
+          typename ExceptionType = rmm::out_of_memory>
```

Comment:
*issue*: We are moving away from having templated adaptors and instead just using type-erased resource_ref objects (see https://github.com/rapidsai/rmm/pull/1661). To avoid introducing a new such class here, can we immediately move to the new model:

```
template<typename ExceptionType = rmm::out_of_memory>
class failure_alternate_resource_adaptor final : public device_memory_resource {
   ...
   failure_alternate_resource_adaptor(device_async_resource_ref primary, device_async_resource_ref alternate) : ... {}
   
```

Or, if we can't type-erase yet, we should at least accept resource refs as an alternate constructor, and store resource_refs rather than templated type-specific usptreams.

Reply:
Done in cbd7f43d5093438eaa6d0fde4a3f5fe7f1147f24, which  rely on implicit type conversion from `device_memory_resource*`  to `device_async_resource_ref`